### PR TITLE
fix: Recreate Analytics clients when credentials change

### PR DESCRIPTION
**Pull request by jimblanc**
#### Description of changes

This change fixes an issue where Analytics clients for Kinesis, Personalize, and Kinesis Firehose weren't being updated when credentials change (e.g. after expiration). Pinpoint is not impacted by this issue as it uses the legacy buffer that specifically handles this case.

#### Description of how you validated changes
Tested providers in sample applications by waiting for credentials to expire and verifying that new events can be recorded. E2E test suite run which covers Pinpoint, Personalize, and Kinesis.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/analytics/src/providers/kinesis-firehose/utils/getEventBuffer.ts [hunk 1]
```diff
@@ -23,7 +29,7 @@ const eventBufferMap: Record<
 	string,
 	EventBuffer<KinesisFirehoseBufferEvent>
 > = {};
-const cachedClients: Record<string, FirehoseClient> = {};
+const cachedClients: Record<string, [FirehoseClient, AWSCredentials]> = {};
```
Comment (jimblanc) [inline packages/analytics/src/providers/kinesis-firehose/utils/getEventBuffer.ts:32]: Need to cache credentials in-memory alongside the associated client instance in order to determine if they changed. SDK clients do not provide a synchronous method for fetching their configured credentials.

Comment (erinleigh90) [reply packages/analytics/src/providers/kinesis-firehose/utils/getEventBuffer.ts:32]: Makes sense to me

**Review by erinleigh90 — DISMISSED** [review]
LGTM!

**Review by cshfang — COMMENTED** [review]
General comment - is it at all possible (and what are the tradeoffs if so) to simply not cache clients at all?

Also seems kind of evident we could benefit from making these providers a bit more DRY to avoid having to modify the same thing in multiple places but that's probably outside of the scope of this PR

**Comment by jimblanc** [comment]
> General comment - is it at all possible (and what are the tradeoffs if so) to simply not cache clients at all?
> 
> Also seems kind of evident we could benefit from making these providers a bit more DRY to avoid having to modify the same thing in multiple places but that's probably outside of the scope of this PR

We could avoid caching any clients by just creating a new one each time we have events to send, but I think the performance & memory benefits from not creating/destroying clients for each batch make it worthwhile. Agree with more DRY, but out of scope for sev-2 (and to reduce the risk surface of the fix).

File: packages/analytics/src/providers/kinesis-firehose/utils/getEventBuffer.ts [hunk 2]
```diff
@@ -76,18 +82,28 @@ export const getEventBuffer = ({
 	userAgentValue,
 }: KinesisFirehoseEventBufferConfig): EventBuffer<KinesisFirehoseBufferEvent> => {
 	const sessionIdentityKey = [region, identityId].filter(id => !!id).join('-');
+	const cachedClient = cachedClients[sessionIdentityKey];
```
[the remark this answers is not in the document]
Comment (jimblanc) [reply packages/analytics/src/providers/kinesis-firehose/utils/getEventBuffer.ts:85]: > Does this need to be tied to per-client? I.e. is it possible for one cached client to become stale due to credential changes without the others to also?

Potentially, the reason that these maps exist is to ensure that all events are transmitted if the identity ID changes, which may have different sets of credentials.

File: packages/analytics/src/providers/kinesis/utils/getEventBuffer.ts [hunk 3]
```diff
@@ -67,19 +73,29 @@ export const getEventBuffer = ({
...
-	if (!eventBufferMap[sessionIdentityKey]) {
+	if (!eventBufferMap[sessionIdentityKey] || credentialsHaveChanged) {
 		const getKinesisClient = (): IAnalyticsClient<KinesisBufferEvent> => {
-			if (!cachedClients[sessionIdentityKey]) {
-				cachedClients[sessionIdentityKey] = new KinesisClient({
-					credentials,
-					region,
-					customUserAgent: userAgentValue,
-				});
+			if (!cachedClient || credentialsHaveChanged) {
```
[the remark this answers is not in the document]
Comment (jimblanc) [reply packages/analytics/src/providers/kinesis/utils/getEventBuffer.ts:86]: Should be able to use the destructured client ref, updated the other providers to be consistent

File: packages/analytics/src/providers/kinesis-firehose/utils/getEventBuffer.ts (hunk 1 above)
Comment (jimblanc) [reply packages/analytics/src/providers/kinesis-firehose/utils/getEventBuffer.ts:34]: Commented in thread below

File: packages/analytics/src/providers/kinesis-firehose/utils/getEventBuffer.ts [hunk 4]
```diff
@@ -82,17 +84,17 @@ export const getEventBuffer = ({
 	userAgentValue,
 }: KinesisFirehoseEventBufferConfig): EventBuffer<KinesisFirehoseBufferEvent> => {
 	const sessionIdentityKey = [region, identityId].filter(id => !!id).join('-');
-	const cachedClient = cachedClients[sessionIdentityKey];
+	const [ cachedClient, cachedCredentials ] = cachedClients[sessionIdentityKey] ?? [];
```
Comment (erinleigh90) [inline packages/analytics/src/providers/kinesis-firehose/utils/getEventBuffer.ts:87]: Much easier to read!

**Review by erinleigh90 — DISMISSED** [review]
Looks even better to me!

File: packages/analytics/src/providers/kinesis-firehose/utils/getEventBuffer.ts [hunk 5]
```diff
@@ -76,18 +84,28 @@ export const getEventBuffer = ({
...
+						region,
+						credentials,
+						customUserAgent: userAgentValue,
+					}),
+					credentials
+				];
 			}
 
-			const firehoseClient = cachedClients[sessionIdentityKey];
+			const firehoseClient = cachedClients[sessionIdentityKey][0];
```
Comment (cshfang) [inline packages/analytics/src/providers/kinesis-firehose/utils/getEventBuffer.ts:108]: Clarity nit again:
```suggestion
			const [ firehoseClient ] = cachedClients[sessionIdentityKey];
```

**Review by cshfang — DISMISSED** [review]
Just a nit but I think this lgtm

**Review by cshfang — APPROVED** [review]

**Review by erinleigh90 — APPROVED** [review]

**Pull request merged**